rfhub2/cli/population.py

```diff
@@ -0,0 +1,175 @@
+from rfhub2.config import APP_INTERFACE, APP_PORT, BASIC_AUTH_USER, BASIC_AUTH_PASSWORD
+import os
+import re
+
+RESOURCE_PATTERNS = {".robot", ".txt", ".tsv", ".resource"}
+ALL_PATTERNS = (RESOURCE_PATTERNS | {".xml", ".py"})
+PROTOCOL = 'http://'
+API_V1 = 'api/v1'
+
+
+class LibraryPopulation(object):
+
+    def __init__(self, paths, no_installed_keywords):
+        self.paths = paths
+        self.auth = (BASIC_AUTH_USER, BASIC_AUTH_PASSWORD)
+        self.no_installed_keywords = no_installed_keywords
+
+    def add_collections(self) -> None:
+        """
+        Traverses through paths and adds libraries to rfhub.
+        :return:
+        """
+
+        def traverse_paths(path):
+            for item in os.listdir(path):
+                full_path = os.path.join(path, item)
+                if os.path.isdir(full_path):
+                    if self._is_library_with_init(full_path):
+                        self.add(full_path)
+                    else:
+                        traverse_paths(full_path)
+                elif os.path.isfile(full_path) and self._is_robot_file(full_path):
+                    self.add(full_path)
+
+        for path in self.paths:
+            traverse_paths(path)
+        if not self.no_installed_keywords:
+            libdir = os.path.dirname(robot.libraries.__file__)
+            for item in os.listdir(libdir):
+                if not self._should_ignore(item):
+                    self.add(os.path.join(libdir, item))
+
+    def add(self, path) -> None:
+        """
+        Adds library with keywords to rfhub.
+        :return:
+        """
+
+        def _serialise_libdoc() -> None:
+            """
+            Serialises libdoc object to dict object.
+            :param libdoc: libdoc input object
+            :param path: library path
+            :return: json object with parameters needed for request post method
+            """
+
+            def _libdoc_to_dict():
+                lib_dict = libdoc.__dict__
+                lib_dict['doc_format'] = lib_dict.pop('_setter__doc_format')
+                for key in ('_setter__keywords', 'inits', 'named_args'):
+                    lib_dict.pop(key)
+                return lib_dict
+
+            lib_dict = _libdoc_to_dict()
+            lib_dict['path'] = path
+            return lib_dict
+
+        def _serialise_keywords():
+            keywords = [keyword.__dict__ for keyword in libdoc.keywords]
+            for keyword in keywords:
+                keyword.pop('tags')
+                if keyword["args"]:
+                    keyword["args"] = str([str(item) for item in keyword["args"]]).replace("\'", "\"")
+                else:
+                    keyword["args"] = ""
+            return keywords
+
+        libdoc = LibraryDocumentation(path)
+        serialised_keywords = _serialise_keywords()
+        serialised_libdoc = _serialise_libdoc()
+        s = session()
+        coll_req = self._post_request(s, 'collections', serialised_libdoc)
```

Comment:
I think it would be nice if it would be refactor in such a way that ApiClient method would both take as an argument and return instances of model classes, so you would not have any request/json related logic here

Reply:
Can You take a look, at changes I did? I'm still not sure about it.